snapshot on scale to zero and shutdown

avinassh opened the pull request:
closes https://github.com/libsql/sqld/issues/757

- I renamed `Database.shutdown` to `Database.destroy`, which sends only the close signal. The new `Database.shutdown` now sends the close signal + waits for snapshot to complete

avinassh commented on libsql-server/src/lib.rs:
```diff
@@ -365,6 +365,7 @@ where
         let snapshot_callback = self.make_snapshot_callback();
         let auth = self.user_api_config.get_auth().map(Arc::new)?;
         let extensions = self.db_config.validate_extensions()?;
+        let mut primary_namespace_store: Option<NamespaceStore<PrimaryNamespaceMaker>> = None;
```
I couldn't think of a better place than here to keep the NamespaceStore reference

MarinPostma replied:
There is another way you can do that, which doesn't leak the primary implementation details to a wider scope: you put it in a future. You rename this to `namespace_store_shutdown_fut: Pin<Box<dyn Future<Output = Result<()>>>>`, and declare it as:
```rust
namespace_store_shutdown_fut = Box::pin(async move {
    namespace_store.shutdown().await
})
```

in both the replica and primary branch :)

The advantage of that is that if we decide  to extend the shutdown method for replica, everything is already in place to do that

MarinPostma commented on libsql-server/src/namespace/mod.rs:
```diff
@@ -495,6 +497,15 @@ impl<M: MakeNamespace> NamespaceStore<M> {
         Ok(())
     }
 
+    pub async fn shutdown(&self) -> crate::Result<()> {
```
you should mark the store as shut down to prevent other namespaces to be reallocated after this is called

avinassh replied:
Could you elaborate? Do you simply mean keep some variable like `has_shutdown`, set it to true at shutdown and use this boolean at other places while creating/forking namespace?

MarinPostma replied:
you won't be able to take ownership because you only have a ref, so yes, the first option sounds great :)

MarinPostma commented on libsql-server/src/replication/primary/logger.rs:
```diff
@@ -266,13 +273,16 @@ unsafe impl WalHook for ReplicationLoggerHook {
-                {
-                    tracing::error!("Failed to snapshot the main db file during checkpoint: {e}");
-                    return SQLITE_IOERR_WRITE;
+                if let Some(replicator) = replicator.lock().unwrap().as_mut() {
```
We want to abort the operation if we find out that the replicator is not there anymore, by returning an error. Same in all cases where we check for the replicator, including in `on_frame`

avinassh commented:
@MarinPostma I have addressed the review comments. I have run into a bug. Here is what I did: 

1. Connected to a DB using `turso db shell`
2. I added a sleep statement right after shutdown future is called - https://github.com/tursodatabase/libsql/blob/d07e3f9/libsql-server/src/lib.rs#L447 
3. Shutdown the server and when it was sleeping, I tried to open a new DB, which failed as expected
4. However the old connection (shell opened earlier) still accepted new reads / writes 

How do I prevent this? I would like to close all connections before calling `namespace_store_shutdown_fut`

MarinPostma commented:
@avinassh in the implementation of `on_frames` in the logger, you want to check that the replicator is still there before doing anything: https://github.com/tursodatabase/libsql/blob/main/libsql-server/src/replication/primary/logger.rs#L81

avinassh commented on libsql-server/src/replication/primary/logger.rs:
```diff
@@ -84,6 +85,21 @@ unsafe impl WalHook for ReplicationLoggerHook {
+            }
+        } else {
+            replicator_mutex = None;
+        }
```
in `on_frame`, I wanted to check if bottomless is enabled and if so, hold the lock till the operation is done. Let me know if there is a cleaner way to do above

LucioFranco commented on libsql-server/src/database.rs:
```diff
@@ -35,14 +47,26 @@ pub struct PrimaryDatabase {
+        self.logger.closed_signal.send_replace(true);
+    }
+
+    async fn shutdown(self) -> Result<()> {
         self.logger.closed_signal.send_replace(true);
+        if let Some(replicator) = &self.logger.bottomless_replicator {
+            let replicator = replicator.lock().unwrap().take();
+            if let Some(mut replicator) = replicator {
+                replicator.wait_until_snapshotted().await?;
+            }
+        }
+        Ok(())
```
Is this cancel safe? @MarinPostma do we have shutdown timeouts? What happens if this gets canceled? Do we have a half backed up backup? This is just waiting so I assume the actual backup happens on a bg thread that can't get cancelled while doing IO by tokio?

MarinPostma replied:
@Horusiath ☝️

MarinPostma replied:
afaik, there are recovery mecanisms when we detect primary is more up to date than bottomless

avinassh replied:
> Is this cancel safe? @MarinPostma do we have shutdown timeouts? What happens if this gets canceled? Do we have a half backed up backup? This is just waiting so I assume the actual backup happens on a bg thread that can't get cancelled while doing IO by tokio?

In unusual circumstances, it can get cancelled in the middle of a shutdown.. may be cos of OOM or from Fly. I had tested this scenario earlier, IIRC, when the machine reboots, restoration from the last back up fails. This is easy to test via Nemesis, I will do that and report back.

MarinPostma commented on libsql-server/src/lib.rs:
```diff
@@ -456,6 +456,7 @@ where
                 namespace_store_shutdown_fut.await?;
                 // clean shutdown, remove sentinel file
                 std::fs::remove_file(sentinel_file_path(&self.path))?;
+                println!("server shutdown complete -- clean exit");
```
please use `tracing::info!` for that

avinassh replied:
hey yes, I have reverted it to `trace`. I wanted to check if the tests time out, so I pushed something quickly. 

@LucioFranco I think the timeout setting is not working as expected. There is a deadlock in my PR and this particular `test::bottomless::rollback_restore` test never finishes. However, its been more than 10 mins and CI hasn't cancelled it.

avinassh commented:
>  in the implementation of on_frames in the logger, you want to check that the replicator is still there before doing anything

Yes, I did this now. However, the issue was happening because I missed checking `has_shutdown` in `with` / `authenticated` methods and it was initiating new bottomless by using `try_new_primary` 

I have addressed all the review comments. But the one test `test::bottomless::rollback_restore` never finishes. The last log I see is 

```
[2023-11-10T15:08:21Z DEBUG bottomless::replicator] Uploaded to S3: testbackuprestore/ns-testbackuprestore:default-e4e4c070-f7bf-7097-961c-a98e1312c48b/000000001157-000000002000-1699628900.gz in 3.487429ms
[2023-11-10T15:08:22Z TRACE bottomless::replicator] Requesting flush
[2023-11-10T15:08:22Z TRACE bottomless::replicator] Confirmed commit of frame no. 2000 (waited for >= 2000)
[2023-11-10T15:08:22Z DEBUG bottomless::replicator] waiting for generation snapshot e4e4c070-f7bf-7097-961c-a98e1312c48b to complete
```

I am suspecting I introduced a deadlock somewhere. Do you see anything obvious in my PR?

LucioFranco commented on libsql-server/src/namespace/mod.rs:
```diff
@@ -348,6 +350,7 @@ struct NamespaceStoreInner<M: MakeNamespace> {
     /// The namespace factory, to create new namespaces.
     make_namespace: M,
     allow_lazy_creation: bool,
+    has_shutdown: RwLock<bool>,
```
Why is this a RwLock and not a Mutex? And infact we can probably use AtomicBool instead?

avinassh replied:
> and Ill add to this why is it an async RwLock and not a sync one?

I didn't give much thought to it. The existing code used async lock, since the module had async operations, I used async lock. `has_shutdown`, we can use sync lock/mutex. Should we always default to using sync mutex if that works?

> Why is this a RwLock and not a Mutex? 
The value would be mutated only once, rest all are read operations. So I thought RWLock would be better 

> And infact we can probably use AtomicBool instead?
I guess we can

MarinPostma replied:
yes, an `AtomicBool` is sufficient

MarinPostma commented on libsql-server/src/replication/primary/logger.rs:
```diff
@@ -84,6 +85,21 @@ unsafe impl WalHook for ReplicationLoggerHook {
+            replicator_lock = replicator.lock().unwrap();
+            if replicator_lock.is_none() {
+                tracing::error!("fatal error: expected a replicator, exiting");
+                std::process::abort()
```
do not abort the process here, just return an io error instead

MarinPostma replied:
we just need to prevent the write, no need to shutdown the whole instance

avinassh replied:
Thats a big oops from my side. I had done this at four places and replaced all of them to return `SQLITE_IOERR`

MarinPostma commented on libsql-server/src/replication/primary/logger.rs:
```diff
@@ -162,12 +177,16 @@ unsafe impl WalHook for ReplicationLoggerHook {
+                    replicator.rollback_to_frame(last_valid_frame);
+                } else {
+                    tracing::error!("fatal error: expected a replicator, exiting");
+                    std::process::abort()
```
same here
